### PR TITLE
Add UI warnings to MFA recommended accounts

### PR DESCRIPTION
**Pull request by jenshenny**
Builds from and is blocked by https://github.com/rubygems/rubygems.org/pull/2987
Part of: https://github.com/rubygems/rubygems.org/pull/2994

This PR contributes to Phase 2 of the [MFA RFC](https://github.com/rubygems/rfcs/pull/36). More specifically, after login, it redirects users to setup or change their MFA setting if they are recommended to (based on a gem threshold). Users can disregard the warning at this stage.

### 1. User is not under `mfa_recommended?`
no change, redirects to user dashboard.

### 2. User is under `mfa_recommended?` and has mfa enabled on the `ui_and_gem_signin` and `ui_and_api` level
no change, redirects to user dashboard.

### 3. User has mfa disabled and is under `mfa_recommended?`
Redirect to mfa setup page and show the following banner
```
For protection of your account and your gems, we encourage you to set up multifactor authentication.
```
<details>
  <summary>Video</summary>
  
https://user-images.githubusercontent.com/42748004/158090103-157f404e-8af3-4dd2-a7a4-2856f17d9c77.mov

</details>

### 4. User has mfa enabled under `ui_only` and is under `mfa_recommended?`
Redirect to user settings and show the following banner
```
For protection of your account and your gems, we encourage you to change your MFA level to "UI and gem signin" or "UI and API".
```

<details>
  <summary>Video</summary>
 
https://user-images.githubusercontent.com/42748004/158089913-7b773efa-e480-4cf7-be8a-cb67fe6dd645.mov

</details>

### CODE REVIEW

**Comment by simi** [comment]
:partying_face: looks ok to me. The only remaining question is what to do regarding codeclimate report. Should we ignore it, disable it or move the mfa methods to some kind of concern and include into User model? Any prefernce @sonalkr132 ?

**Review by simi — APPROVED** [review]

**Comment by jenshenny** [comment]
> move the mfa methods to some kind of concern and include into User model

The amount of methods related to MFA is growing so I would be ok with moving it and including it into the user (I would prefer do this as a refactor after we get the warnings shipped though).

**Comment by simi** [comment]
sounds like a plan

**Comment by sonalkr132** [comment]
Do we want to mention in the message that MFA will be enforced in near future? Its okay if we want to do it later. I feel at some point we should mention it and perhaps link the RFC/blog.

**Comment by sonalkr132** [comment]
> I would prefer do this as a refactor after we get the warnings shipped though

Please add the file to exclusion https://docs.codeclimate.com/docs/excluding-files-and-folders and create an issue for it.

**Comment by jenshenny** [comment]
> Do we want to mention in the message that MFA will be enforced in near future? Its okay if we want to do it later. I feel at some point we should mention it and perhaps link the RFC/blog.

I added it as a note as a part of the [CLI warnings](https://github.com/Shopify/ruby-conventions/issues/84) so it makes sense to add it here also. I added a commit that does that ✅ 

> Please add the file to exclusion https://docs.codeclimate.com/docs/excluding-files-and-folders and create an issue for it.

I will create an issue. I don't have access to the code climate settings to add a file exclusion however :(

**Comment by simi** [comment]
@jenshenny I think you can just add file to this repo in this PR.

https://docs.codeclimate.com/docs/advanced-configuration using `exclude_patterns` (https://docs.codeclimate.com/docs/advanced-configuration#examples)

**Comment by jenshenny** [comment]
Thanks @simi! It looks like adding `.codeclimate.yml` (with a single file exclusion) overrides any in app settings that are configured. I don't know the configurations made on the app end so it surfaced a bunch of issues. Should we be using a `.codeclimate.yml` for these configurations instead of relying on the UI at some point?

Screen Shot 2022-04-11 at 3 08 28 PM

**Comment by jenshenny** [comment]
Adding those defaults did solve the issues 🙌

**Comment by simi** [comment]
Code wise it seems ok to me and CI seems happy as well. I'm not sure about the text part. Would it be possible to also explain why your account is subject to get enforced MFA later (related to gems download)? Should we create a page at https://guides.rubygems.org/security/ before release of this feature and link in there for more details? Is there any plan on this?

**Comment by jenshenny** [comment]
We have a draft PR regarding the blog post planned to be sent out https://github.com/rubygems/rubygems.github.io/pull/110. As @sonalkr132 [mentioned](https://github.com/rubygems/rubygems.org/pull/2994#issuecomment-1095358478), it seems like it would be a good call to link that information to the text. wdyt? Adding a page to the guides would also be good for visibility. 

> before release of this feature and link in there for more details? Is there any plan on this?

That is true... I think we should merge this PR, https://github.com/rubygems/rubygems.org/pull/2998, and any communications (eg. https://github.com/rubygems/rubygems.github.io/pull/110) at the same time.

File: test/functional/sessions_controller_test.rb [hunk 1]
```diff
@@ -5,7 +5,6 @@ class SessionsControllerTest < ActionController::TestCase
     setup do
       @user = User.new(email_confirmed: true, handle: "test")
       @user.enable_mfa!(ROTP::Base32.random_base32, :ui_only)
-      @request.cookies[:mfa_feature] = "true"
```
Comment (aellispierce) [inline test/functional/sessions_controller_test.rb:8]: This isn't strictly relevant to this change, I just noticed that this unused flag was still kicking around in tests and thought this a prudent time to remove it.

**Comment by aellispierce** [comment]
@sonalkr132 As we discussed, I've added a cookie feature-flag here so that this can be deployed and tested in production. Would you mind taking a look at the [last commit](https://github.com/rubygems/rubygems.org/commit/0e1d84cbdae3bcff0aa8d2ea463517a9c17149cc) to make sure this is still 👍🏾 ?

**Review by sonalkr132 — APPROVED** [review]
LGTM. thank you :100:

**Pull request merged**